File: arrays backup/Sortings/quick sort/quick sort.cpp
#include <iostream>


void swap(int *array, int a, int b) {
    int temp = array[a];
    array[a] = array[b];
    array[b] = temp;
}

int partition(int *array, int begin, int end) {
    int pivot = array[end];
    int i = begin - 1;
    for (int j = begin; j < end; j++)  
        if (array[j] <= pivot) 
            swap(array, j, ++i);
    
    swap(array, end, ++i);
    return i;
}

void quick_sort(int *array, int begin, int end) {
    if (begin < end) {
        int i = partition(array, begin, end);
        quick_sort(array, begin, i-1);
        quick_sort(array, i+1, end);
    }
}

int main() {
    int len = 20;
    int *array = new int[len];

    // fill with random elements
    srand(time(NULL));
    for (int i = 0; i < len; i++) 
        array[i] = rand() % 30;

    // print
    for (int i = 0; i < len; i++) 
        std::cout << array[i] << " ";

    std::cout << "\n";
    quick_sort(array, 0, len-1);

    // print
    for (int i = 0; i < len; i++) 
        std::cout << array[i] << " ";

    delete array;

    return 0;
}

File: arrays backup/Sortings/quick sort/quick_sort.py
def partition(arr, begin, end):
    pivot = arr[end]
    i = begin - 1  # smaller element index (bottom)
    for j in range(begin, end):  # greater element index
        if arr[j] <= pivot:
            i += 1
            arr[i], arr[j] = arr[j], arr[i]  # swap
    # pivot <> last smaller element (bottom)
    arr[i+1], arr[end] = arr[end], arr[i+1]
    return i + 1

def quick_sort(arr, begin, end):
    if begin < end:
        # index that divides array (smaller num index)
        i = partition(arr, begin, end)
        quick_sort(arr, begin, i - 1)  # sorting before pivot
        quick_sort(arr, i + 1, end)    # sorting after pivot
        return arr

# Driver code
#to_sort = [0, 5, 87, 12, 24, 78, 23, 67, 12]
#print(quick_sort(to_sort, 0, len(to_sort)-1))
